packages/shared-lib/src/lib/JobQueueWithClasses.ts

```diff
@@ -93,41 +94,53 @@ export class JobQueueWithClasses {
 	/** Start executing the queue */
 	start(): void {
 		if (this.#paused === false) return
-		;(async () => {
-			this.#paused = false
-			// eslint-disable-next-line no-constant-condition
-			while (true) {
-				const firstIn = this.#queue.shift()
-				if (!firstIn) {
-					// No more jobs on queue.
-					this.#waitForDonePromises.forEach((p) => {
-						p.resolve()
-					})
-					break
-				}
-				try {
-					await firstIn.fn()
-					firstIn.resolve()
-				} catch (error) {
-					firstIn.reject(error)
+		Promise.resolve()
+			.then(async () => {
+				this.#paused = false
+				// eslint-disable-next-line no-constant-condition
+				while (true) {
+					const firstIn = this.#queue.shift()
+					if (!firstIn) {
+						// No more jobs on queue.
+						if (this.#waitForDonePromise) {
+							this.#waitForDonePromise.resolve()
+							this.#waitForDonePromise = undefined
+						}
+						break
+					}
+					try {
+						this.#jobIsRunning = true
+						await firstIn.fn()
+						this.#jobIsRunning = false
+						firstIn.resolve()
+					} catch (error) {
+						this.#jobIsRunning = false
+						firstIn.reject(error)
+					}
 				}
-			}
-			this.#paused = true
-		})().catch((error) => {
-			console.error(error)
-		})
+				this.#paused = true
+			})
+			.catch((error) => {
+				console.error(error)
+			})
 	}
 	/** Returns the count of waiting jobs in the queue (ie not started yet) */
 	getWaiting(): number {
 		return this.#queue.length
 	}
-	/** Returns a Promise that resolves when the queue is eventually empty */
+	/** Returns a Promise that resolves when the queue is eventually empty and all jobs are done */
 	async waitForDone(): Promise<void> {
```

Comment:
I think this `waitForDone` implementation has a race condition.

If you call `add()`, don't call `start()` (because autostart was set to true), then immediately call `waitForDone()`, `this.#jobIsRunning` will be false. as `start()` hasnt been run yet (its inside of a `setImmediate`), resulting in the call to `waitForDone()` returning `Promise.resolve()` instead of creating the correct promise. 
I think that dropping the ` && !this.#jobIsRunning` from L133, I don't see the purpose in that being part of the guard